File: solutions/typescript/split-second-stopwatch/1/split-second-stopwatch.ts
type StopwatchState = 'ready' | 'running' | 'stopped'

export class SplitSecondStopwatch {
  private _state: StopwatchState = 'ready'
  private currentLapSeconds = 0
  private previousLapSeconds: number[] = []

  private static parseDuration(duration: string): number {
    const [hours, minutes, seconds] = duration.split(':').map(Number)
    return hours * 3600 + minutes * 60 + seconds
  }

  private static formatDuration(totalSeconds: number): string {
    const hours = Math.floor(totalSeconds / 3600)
    const minutes = Math.floor((totalSeconds % 3600) / 60)
    const seconds = totalSeconds % 60

    return [
      String(hours).padStart(2, '0'),
      String(minutes).padStart(2, '0'),
      String(seconds).padStart(2, '0'),
    ].join(':')
  }

  public get state(): string {
    return this._state
  }

  public get currentLap(): string {
    return SplitSecondStopwatch.formatDuration(this.currentLapSeconds)
  }

  public get total(): string {
    const totalSeconds =
      this.previousLapSeconds.reduce(
        (sum, lap) => sum + lap,
        0,
      ) + this.currentLapSeconds

    return SplitSecondStopwatch.formatDuration(totalSeconds)
  }

  public get previousLaps(): string[] {
    return this.previousLapSeconds.map(
      SplitSecondStopwatch.formatDuration,
    )
  }

  public start(): void {
    if (this._state === 'running') {
      throw new Error(
        'cannot start an already running stopwatch',
      )
    }

    this._state = 'running'
  }

  public stop(): void {
    if (this._state !== 'running') {
      throw new Error(
        'cannot stop a stopwatch that is not running',
      )
    }

    this._state = 'stopped'
  }

  public lap(): void {
    if (this._state !== 'running') {
      throw new Error(
        'cannot lap a stopwatch that is not running',
      )
    }

    this.previousLapSeconds.push(this.currentLapSeconds)
    this.currentLapSeconds = 0
  }

  public reset(): void {
    if (this._state !== 'stopped') {
      throw new Error(
        'cannot reset a stopwatch that is not stopped',
      )
    }

    this._state = 'ready'
    this.currentLapSeconds = 0
    this.previousLapSeconds = []
  }

  public advanceTime(duration: unknown): void {
    if (this._state !== 'running') {
      return
    }

    this.currentLapSeconds +=
      SplitSecondStopwatch.parseDuration(
        duration as string,
      )
  }
}

File: solutions/go/split-second-stopwatch/1/split_second_stopwatch.go
package splitsecondstopwatch

import "fmt"

type SplitSecondStopwatch struct {
	state        string
	currentLap   int64
	total        int64
	previousLaps []int64
}

func formattedTime(time int64) string {
	hours := time / 3600000
	minutes := (time % 3600000) / 60000
	seconds := (time % 60000) / 1000

	return fmt.Sprintf("%02d:%02d:%02d", hours, minutes, seconds)
}

func timeFromString(s string) int64 {
	var hours, minutes, seconds int64
	fmt.Sscanf(s, "%02d:%02d:%02d", &hours, &minutes, &seconds)
	return hours*3600000 + minutes*60000 + seconds*1000
}

func (sss *SplitSecondStopwatch) Start() error {
	if sss.state == "running" {
		return fmt.Errorf("cannot start an already running stopwatch")
	}

	sss.state = "running"

	return nil
}

func (sss *SplitSecondStopwatch) Stop() error {
	if sss.state != "running" {
		return fmt.Errorf("cannot stop a stopwatch that is not running")
	}

	sss.state = "stopped"

	return nil
}

func (sss *SplitSecondStopwatch) Reset() error {
	if sss.state == "ready" || sss.state == "running" {
		return fmt.Errorf("cannot reset a stopwatch that is not stopped")
	}

	sss.state = "ready"
	sss.total = 0
	sss.currentLap = 0
	sss.previousLaps = nil

	return nil
}

func (sss *SplitSecondStopwatch) Lap() error {
	if sss.state != "running" {
		return fmt.Errorf("cannot lap a stopwatch that is not running")
	}

	sss.previousLaps = append(sss.previousLaps, sss.currentLap)
	sss.currentLap = 0

	return nil
}

func (sss *SplitSecondStopwatch) AdvanceTime(by string) {
	if sss.state != "running" {
		return
	}

	time := timeFromString(by)
	sss.currentLap += time
	sss.total += time
}

func (sss *SplitSecondStopwatch) State() string {
	return sss.state
}

func (sss *SplitSecondStopwatch) CurrentLap() string {
	return formattedTime(sss.currentLap)
}

func (sss *SplitSecondStopwatch) Total() string {
	return formattedTime(sss.total)
}

func (sss *SplitSecondStopwatch) PreviousLaps() []string {
	laps := make([]string, len(sss.previousLaps))
	for i, lap := range sss.previousLaps {
		laps[i] = formattedTime(lap)
	}
	return laps
}

func NewSplitSecondStopwatch() *SplitSecondStopwatch {
	return &SplitSecondStopwatch{state: "ready"}
}